src/graph/transform/cpu/knn.cc

```diff
@@ -15,6 +20,221 @@ namespace dgl {
 namespace transform {
 namespace impl {
 
+// This value is directly from pynndescent
+static constexpr int NN_DESCENT_BLOCK_SIZE = 16384;
+
+/*!
+ * \brief Compute Euclidean distance between two vectors, return positive
+ *  infinite value if the intermediate distance is greater than the worst
+ *  distance.
+ */
+template <typename FloatType, typename IdType>
+FloatType EuclideanDistWithCheck(const FloatType* vec1, const FloatType* vec2, int64_t dim,
+                                 FloatType worst_dist = std::numeric_limits<FloatType>::max()) {
+  FloatType dist = 0;
+  bool early_stop = false;
+  IdType idx = 0;
+  while (idx < dim - 3) {
+    FloatType diff0 = vec1[idx] - vec2[idx];
+    FloatType diff1 = vec1[idx + 1] - vec2[idx + 1];
+    FloatType diff2 = vec1[idx + 2] - vec2[idx + 2];
+    FloatType diff3 = vec1[idx + 3] - vec2[idx + 3];
+    dist += diff0 * diff0 + diff1 * diff1 + diff2 * diff2 + diff3 * diff3;
+    idx += 4;
+    if (dist > worst_dist) {
+      early_stop = true;
+      idx = dim;
+      break;
+    }
+  }
+
+  // last 3 elements
+  while (idx < dim) {
+    FloatType diff = vec1[idx] - vec2[idx];
+    dist += diff * diff;
+    ++idx;
+    if (dist > worst_dist) {
+      early_stop = true;
+      break;
+    }
+  }
+
+  if (early_stop) {
+    return std::numeric_limits<FloatType>::max();
+  } else {
+    return dist;
+  }
+}
+
+/*! \brief Compute Euclidean distance between two vectors */
+template <typename FloatType, typename IdType>
+FloatType EuclideanDist(const FloatType* vec1, const FloatType* vec2, int64_t dim) {
+  FloatType dist = 0;
+  IdType idx = 0;
+  while (idx < dim - 3) {
+    FloatType diff0 = vec1[idx] - vec2[idx];
+    FloatType diff1 = vec1[idx + 1] - vec2[idx + 1];
+    FloatType diff2 = vec1[idx + 2] - vec2[idx + 2];
+    FloatType diff3 = vec1[idx + 3] - vec2[idx + 3];
+    dist += diff0 * diff0 + diff1 * diff1 + diff2 * diff2 + diff3 * diff3;
+    idx += 4;
+  }
+
+  // last 3 elements
+  while (idx < dim) {
+    FloatType diff = vec1[idx] - vec2[idx];
+    dist += diff * diff;
+    ++idx;
+  }
+
+  return dist;
+}
+
+/*! \brief Insert a new element into a heap */
+template <typename FloatType, typename IdType>
+void HeapInsert(IdType* out, FloatType* dist,
```

Comment:
Also the same question: when to use which?

Reply:
If we want a heap without duplicated elements (we may try to insert a node into this heap multiple times since there may be multiple 2-hop paths from the current node to `x`), we use the checked version. If we are sure that an element will never be inserted twice, we use `HeapInsert` without check.